api/v1/porterconfig_types_test.go

```diff
@@ -40,33 +39,118 @@ func TestPorterConfigSpec_MergeConfig(t *testing.T) {
-		Debug:                pointer.BoolPtr(true),
-		DebugPlugins:         pointer.BoolPtr(true),
-		Namespace:            pointer.StringPtr("test"),
-		Experimental:         []string{"build-drivers"},
-		BuildDriver:          pointer.StringPtr("buildkit"),
-		DefaultStorage:       pointer.StringPtr("in-cluster-mongodb"),
-		DefaultSecrets:       pointer.StringPtr("keyvault"),
-		DefaultStoragePlugin: pointer.StringPtr("mongodb"),
-		DefaultSecretsPlugin: pointer.StringPtr("kubernetes.secrets"),
-		Storage: []StorageConfig{
-			{PluginConfig{
-				Name:         "in-cluster-mongodb",
-				PluginSubKey: "mongodb",
-				Config:       runtime.RawExtension{Raw: []byte(`{"url":"mongodb://..."}`)},
-			}},
+	tests := []struct {
+		name        string
+		cfg         PorterConfigSpec
+		expDocument []byte
+	}{
+		{
+			name: "All fields set",
+			cfg: PorterConfigSpec{
+				Debug:                pointer.BoolPtr(true),
+				DebugPlugins:         pointer.BoolPtr(true),
+				Namespace:            pointer.StringPtr("test"),
+				Experimental:         []string{"build-drivers"},
+				BuildDriver:          pointer.StringPtr("buildkit"),
+				DefaultStorage:       pointer.StringPtr("in-cluster-mongodb"),
+				DefaultSecrets:       pointer.StringPtr("keyvault"),
+				DefaultStoragePlugin: pointer.StringPtr("mongodb"),
+				DefaultSecretsPlugin: pointer.StringPtr("kubernetes.secrets"),
+				Storage: []StorageConfig{
+					{PluginConfig{
+						Name:         "in-cluster-mongodb",
+						PluginSubKey: "mongodb",
+						Config:       runtime.RawExtension{Raw: []byte(`{"url":"mongodb://..."}`)},
+					}},
+				},
+				Secrets: []SecretsConfig{
+					{PluginConfig{
+						Name:         "keyvault",
+						PluginSubKey: "azure.keyvault",
+						Config:       runtime.RawExtension{Raw: []byte(`{"vault": "mysecrets"}`)},
+					}},
+				},
+			},
+			expDocument: []byte(`debug: true
+debug-plugins: true
+namespace: test
+experimental:
+    - build-drivers
+build-driver: buildkit
+default-storage: in-cluster-mongodb
+default-secrets: keyvault
+default-storage-plugin: mongodb
+default-secrets-plugin: kubernetes.secrets
+storage:
+    - config:
+        url: mongodb://...
+      name: in-cluster-mongodb
+      plugin: mongodb
+secrets:
+    - config:
+        vault: mysecrets
+      name: keyvault
+      plugin: azure.keyvault
+`),
 		},
-		Secrets: []SecretsConfig{
-			{PluginConfig{
-				Name:         "keyvault",
-				PluginSubKey: "azure.keyvault",
-				Config:       runtime.RawExtension{Raw: []byte(`{"vault": "mysecrets"}`)},
-			}},
+		{
+			name: "Minimum fields set",
+			cfg: PorterConfigSpec{
+				DefaultStoragePlugin: pointer.StringPtr("mongodb"),
+				DefaultSecretsPlugin: pointer.StringPtr("kubernetes.secrets"),
+			},
+			expDocument: []byte(`default-storage-plugin: mongodb
+default-secrets-plugin: kubernetes.secrets
+`),
+		},
+		{
+			name: "Storage Config minimum fields set",
+			cfg: PorterConfigSpec{
+				DefaultStoragePlugin: pointer.StringPtr("mongodb"),
```

Comment:
You would never create a config like this because DefaultStorage takes precedence over DefaultStoragePlugin. For this test case, I think removing default storage plugin would yield the test case that you are still looking for (leaving the config section empty), no?

I realize we are just testing the schema here, but I think it helps to use examples that could work if used.

Reply:
This is the question I had about "omitempty". For DefaultStoragePlugin "omitempty" is only set for json but not yaml or map structure so not providing this ends up giving you a porter config that looks like this:
`
default-storage: in-cluster-mongodb
default-storage-plugin: null
default-secrets-plugin: kubernetes.secrets
storage:
    - name: in-cluster-mongodb
      plugin: mongodb
`

Reply:
Sorry I didn't catch on from the slack conversation what was wrong! 🤦‍♀️  Yes, the right thing to do is add omitempty for the other formats as well.

Reply:
Ok sounds good, I'll add that. Also DefaultSecretsPlugin doesn't have omitempty for any of the formats, I assume that it should be added to that as well for all